### DISCUSSION TITLE
铁蛋既是铁蛋又是jetpack开发平台的意义

### DISCUSSION THREAD

**Discussion by liseri** [MiRoboticsLab/cyberdog_ros2#45] [Ideas]
问题：
1）感觉铁蛋的jetpack工具不完整；
2）想升级操作系统或jetpack版本，又怕影响到铁蛋正常功能；

述求：
1）铁蛋镜像版本以铁蛋硬件架构为大版本，以os+jetpack版本+ros版本为中版本，其余为小版本的 版本进化；
2）铁蛋有完整的jetpack工具，铁蛋首先是铁蛋其次也是jetpack开发平台；
3）或者镜像可以分为生产版本和开发者版本，铁蛋作为一个成熟的家居产品还不成熟，那么首先抓住发烧友，也是一个不小的市场

**Comment by homalozoa** [comment]
Hi. 感谢您提的建议.

由于项目周期较紧，所以第一代铁蛋在权衡Ubuntu的臃肿和嵌入式Linux的轻量化之间找了个Lubuntu作为权衡，但绝非优解。

1. 目前我们在探索使用Yocto对完整的功能进行打包，也就是诉求3中的生产版本。如果是开发版本，则铁蛋基础之上修改Kernel + BSP基本包 + ROS基本包 + CyberDog_ROS2即可。待我们完全构建完毕后便会发布。
2. 完整的NVIDIA Jetpack工具包，除了BSP相关（如Kernel，Uboot等）被修改外，其余的软件项都是全的，如CUDA Toolkit、Multimedia API，这些可以直接通过apt进行安装。这些软件都是通过https://repo.download.nvidia.com/jetson/ 进行版本管理和获取的，可以参考https://github.com/MiRoboticsLab/cyberdog_ros2/blob/devel/tools/dockers/arm64/Dockerfile#L154 进行了解。
3. 铁蛋由于生产的需求，硬盘的分区表与NVIDIA提供的Sample rootFS无法兼容，且永远无法兼容。NVIDIA提供的Sample rootFS仅是为开发板提供了一个样例，并非用于特定产品，所以不要尝试。

**Comment by liseri** [comment]
@homalozoa  在铁蛋上安装包或更新包的时候经常出现以下log， 不知道是不是必然的
```
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n]
Setting up initramfs-tools (0.130ubuntu3.13) ...
update-initramfs: deferring update (trigger activated)
Processing triggers for flash-kernel (3.98ubuntu11~18.04.2) ...
Unsupported platform.
dpkg: error processing package flash-kernel (--configure):
 installed flash-kernel package post-installation script subprocess returned error exit status 1
Processing triggers for initramfs-tools (0.130ubuntu3.13) ...
update-initramfs: Generating /boot/initrd.img-5.4.0-90-generic
cryptsetup: WARNING: failed to detect canonical device of /dev/root
cryptsetup: WARNING: could not determine root device from /etc/fstab
Warning: root device /dev/root does not exist
Warning: couldn't identify filesystem type for fsck hook, ignoring.
I: The initramfs will attempt to resume from /dev/mmcblk0p13
I: (UUID=1cc44b43-dec3-40ce-abe6-1999e5fb925a)
I: Set the RESUME variable to override this.
/sbin/ldconfig.real: Warning: ignoring configuration file that cannot be opened: /etc/ld.so.conf.d/aarch64-linux-gnu_EGL.conf: No such file or directory
/sbin/ldconfig.real: Warning: ignoring configuration file that cannot be opened: /etc/ld.so.conf.d/aarch64-linux-gnu_GL.conf: No such file or directory
Unsupported platform.
run-parts: /etc/initramfs/post-update.d//flash-kernel exited with return code 1
dpkg: error processing package initramfs-tools (--configure):
 installed initramfs-tools package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 flash-kernel
 initramfs-tools
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Reply by homalozoa** [reply]
> @homalozoa 在铁蛋上安装包或更新包的时候经常出现以下log， 不知道是不是必然的
> 
> ```
> 1 not fully installed or removed.
> After this operation, 0 B of additional disk space will be used.
> Do you want to continue? [Y/n]
> Setting up initramfs-tools (0.130ubuntu3.13) ...
> update-initramfs: deferring update (trigger activated)
> Processing triggers for flash-kernel (3.98ubuntu11~18.04.2) ...
> Unsupported platform.
> dpkg: error processing package flash-kernel (--configure):
>  installed flash-kernel package post-installation script subprocess returned error exit status 1
> Processing triggers for initramfs-tools (0.130ubuntu3.13) ...
> update-initramfs: Generating /boot/initrd.img-5.4.0-90-generic
> cryptsetup: WARNING: failed to detect canonical device of /dev/root
> cryptsetup: WARNING: could not determine root device from /etc/fstab
> Warning: root device /dev/root does not exist
> Warning: couldn't identify filesystem type for fsck hook, ignoring.
> I: The initramfs will attempt to resume from /dev/mmcblk0p13
> I: (UUID=1cc44b43-dec3-40ce-abe6-1999e5fb925a)
> I: Set the RESUME variable to override this.
> /sbin/ldconfig.real: Warning: ignoring configuration file that cannot be opened: /etc/ld.so.conf.d/aarch64-linux-gnu_EGL.conf: No such file or directory
> /sbin/ldconfig.real: Warning: ignoring configuration file that cannot be opened: /etc/ld.so.conf.d/aarch64-linux-gnu_GL.conf: No such file or directory
> Unsupported platform.
> run-parts: /etc/initramfs/post-update.d//flash-kernel exited with return code 1
> dpkg: error processing package initramfs-tools (--configure):
>  installed initramfs-tools package post-installation script subprocess returned error exit status 1
> Errors were encountered while processing:
>  flash-kernel
>  initramfs-tools
> E: Sub-process /usr/bin/dpkg returned an error code (1)
> ```

是因为执行了apt upgrade？

**Reply by liseri** [reply]
真记不清了，但很有可能执行过apt upgrade，尝试过升级到jetpack4.6;  另外我补充下，我之所以说jetpack工具不完整的原因，1)
1)在/usr/local/cuda-10目录下没有bin目录，cuda相关的软链接 环境变量也没有;  2)没有安装nvidia-docker(虽然docker挺耗资源，但可以jetpack有的都装上，然后将耗资源的工具默认不启动即可)

**Reply by homalozoa** [reply]
你的建议很有道理，我们可以维护一个与NVIDIA官方近似的开发版本和一个生产环节的精简版本。我和同事商量一下。

**Comment by Oldpan** [comment]
@liseri @homalozoa 我和也liseri一样尝试更新过铁蛋的 jetpack环境，也遇到了：
```bash
Errors were encountered while processing:
 flash-kernel
 initramfs-tools
```
这些问题，也是怀疑这个ubuntu已经被精简或者修改过，导致我们无法正常按照NVIDIA官方的OTA方式更新。
不过更新后ros2是可以正常跑，铁蛋也可以正常跑，就是cuda无法使用了，显卡驱动不见了，显卡找不到。
想知道一种正确更新铁蛋内部jetpack环境的方式。

另外jetpack工具不完整其实可以通过执行apt install就可以安装那些工具。
我也尝试过在原始版本（jetpack4.5）铁蛋docker pull最新的 jetpack4.6镜像，然后搭建铁蛋环境，不过cuda依然找不到，应该是驱动不兼容的原因，想要更新驱动必须host更新到jetpack4.6才行。

如何正确更新呢，我可以用jetpack的SDK manager方式直接刷机铁蛋，然后再构建铁蛋环境么？或者还有其他的方式？？

**Reply by liseri** [reply]
以我的有限的经验，最省事的办法是按以下步骤完善4.5，4.6非有必要不要整了； 步骤：1）铁蛋刷机，不要upgrade;  2) 使用sdk manager安装4.5; 3)进入铁蛋, update&upgrade;

**Reply by Oldpan** [reply]
哎，我是搞AI的，4.6对我来说作用还挺大，不过目前暂时先不搞了。等熟悉铁蛋内部的环境再说吧。
感谢你的建议，在重刷机好几遍后暂时不打算折腾了。。。目前就先用旧环境，不过是可以apt upgrade的，只是这个之后jtop无法正常检测出jetpack的版本，不过cuda什么的都是可以使用，也就是从4.5升级到了4.5.1

**Reply by liseri** [reply]
jtop确实是出不来，不过不用纠结这个；确实是网上容易查到的资料都是基于标准dev kit的, 工程不是很熟的人，容易在铁蛋上浪费太多时间还不一定能搞定 所以出个标准开发版还是很有意义的 @homalozoa  ; 我买铁蛋的一个理由也是，相当于买个jetson xavier开发板了，算下来还很划算， @Oldpan 你也是这么想的吧 haha

**Reply by Oldpan** [reply]
哈哈，对的 @liseri ，买铁蛋送xavier NX开发板太香了，如果能利用好这块AI性能很强的开发板，铁蛋潜力很大
